control-plane-operator/hostedclusterconfigoperator/controllers/inplaceupgrader/inplaceupgrader.go

```diff
@@ -413,32 +413,43 @@ func getNodesToUpgrade(nodes []*corev1.Node, targetConfig string, maxUnavailable
 	// - NodeNetworkUnavailable condition status is ConditionFalse.
 	// TODO (jerzhang): consider what we want to do with node status here
 	// For now, we will just check current/desired config to see if any nodes is already updating
+
 	var numUnavailable int
+	var candidateNodes []*corev1.Node
 	for _, node := range nodes {
 		if node.Annotations[CurrentMachineConfigAnnotationKey] != node.Annotations[DesiredMachineConfigAnnotationKey] {
 			numUnavailable++
+			// Compared to self-driving OCP, there is an additional scenario to consider here.
+			// Since the ConfigMap contents are synced separately, those will change on the fly in the pod.
+			// Meaning that we could have a scenario where there are multiple queue'ed updates, in which case
+			// the MCD will just jump straight to the latest version.
+			// This will cause the MCD to softlock, so let's make sure for those unavailable nodes, we are also
+			// update their desired configuration. The MCD should be able to reconcile these changes on the fly.
+			if node.Annotations[DesiredMachineConfigAnnotationKey] != targetConfig {
+				candidateNodes = append(candidateNodes, node)
+			}
 		}
 	}
 
 	capacity := maxUnavailable - numUnavailable
 	// If we're at capacity, there's nothing to do.
 	if capacity < 1 {
-		return nil
+		return candidateNodes
```

Comment:
why do we return candidates here, isn't the comment still valid `// If we're at capacity, there's nothing to do.`?

Reply:
So the difference here is that our candidate nodes have two sources:
1. nodes that are currently updating (current != desired) && desired != latest
2. nodes that are currently not updating or updated (current == desired && desired != latest)

So in this case the second would be empty, but we return the list of candidate nodes which is populated from 1, or otherwise empty

Reply:
I'll update the comment here to make that more obvious

Reply:
I see, thanks. As devil starts to show up in the details and this becomes more tricky, is it possible to have some unit tested functions mapping to the higher level idiomatic semantics, e.g something like `getCapacity`, `getAlreadyUnavailableCandidates`, getAvailableCandidates` and also add the test cases intended by this PR to TestGetNodesToUpgradeI?

Reply:
Sure, I can definitely add a couple of new test cases

Reply:
Added unit tests for now to `TestGetNodesToUpgrade`